### PR TITLE
Refactor URL parsing code

### PR DESCRIPTION
**Pull request by kkaefer**
An attempt at https://github.com/mapbox/mapbox-gl-native/issues/6183.

### CODE REVIEW

**Comment by 1ec5** [comment]
It would be great if we could eventually upstream and reconcile this code with https://github.com/mapbox/uri-hpp/.

/cc @mapsam

File: src/mbgl/util/mapbox.cpp [hunk 1]
```diff
@@ -21,188 +23,148 @@ bool isMapboxURL(const std::string& url) {
...
-    }
+    const auto tpl = baseURL + "/v4/{path}?access_token=" + accessToken;
+    return transformURL(tpl, str, url);
+}
 
-    const auto extensionIdx = url.find(".", basenameIdx);
-    if (extensionIdx == std::string::npos || extensionIdx == queryIdx - 1) {
-        return url;
+std::string
+canonicalizeTileURL(const std::string& str, const SourceType type, const uint16_t tileSize) {
```
Comment (jfirebaugh) [inline src/mbgl/util/mapbox.cpp]: The prior version of `canonicalizeTileURL` wasn't restricted to `*.mapbox.com`, right? It definitely needs to continue to support `*.mapbox.cn` (we should add a test for this) and should likely continue to support arbitrary/staging hostnames as well. It's only used in situations where the source URL is `mapbox://...`, so I don't think enforcing a particular hostname is necessary.

**Comment by jfirebaugh** [comment]
👍  on `URL`/`Path` -- seems like a good middle ground between ad hoc parsing and a full blown RFC compliant library.

File: test/util/url.test.cpp [hunk 2]
```diff
@@ -23,3 +23,184 @@ TEST(URL, isURL) {
...
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo.png").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo.png?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 18, 0 }), URLPath("http://example.com").directory);
+    EXPECT_EQ(Path::Segment({ 18, 0 }), URLPath("http://example.com?query=foo.bar").directory);
+}
+
+TEST(Path, Extension) {
```
Comment (jfirebaugh) [inline test/util/url.test.cpp]: Should have a test for double extensions (e.g. `.vector.pbf`).

Comment (kkaefer) [inline test/util/url.test.cpp]: I guess it's debatable on what constitutes an extension and what is part of the filename. For a (hypothetical) URL of `mapbox://sources/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7`, would the extension be nothing? or `.mapbox-terrain-v2,mapbox.mapbox-streets-v7`?

Comment (jfirebaugh) [inline test/util/url.test.cpp]: Looks like you went with "the final period and everything following". Seems good to me!

**Comment by kkaefer** [comment]
For reference, here's an insightful article by the author of cURL on why RFC compliance for URL parsing doesn't work: https://daniel.haxx.se/blog/2016/05/11/my-url-isnt-your-url/

**Pull request merged**